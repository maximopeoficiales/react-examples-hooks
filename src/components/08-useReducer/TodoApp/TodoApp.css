.item{
    cursor: pointer;
}
.complete {

  text-decoration: line-through;
}
