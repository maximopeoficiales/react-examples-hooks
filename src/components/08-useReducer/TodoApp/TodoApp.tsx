import { useEffect, useReducer } from "react";
import TodoForm from "../TodoForm/TodoForm";
import TodoList from "../TodoList/TodoList";
import { Action, Todo, todoReducer } from "../todoReducer";

import "./TodoApp.css";

interface MyProps {}
const defaultProps: MyProps = {};
const TodoApp = (props: MyProps) => {
  props = { ...defaultProps, ...props };

  const init = (): Todo[] => {
    const todos = localStorage.getItem("todos");
    return todos == null ? [] : JSON.parse(todos);
  };
  // tambien se puede usar de esta manera
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const addTodo = (todo: Todo) => {
    dispatch({ type: Action.ADD, payload: todo });
  };
  const deleteTodo = (todo: Todo) => {
    dispatch({ type: Action.DELETE, payload: todo });
  };
  const toogleTodo = (todo: Todo) => {
    dispatch({ type: Action.TOGGLE, payload: todo });
  };

  return (
    <div data-testid="TodoApp" className="row justify-content-center">
      <div className="col-md-10">
        <h1 className="text-center font-weight-bold my-4">TodoApp component</h1>

        <div className="row">
          <div className="col-md-7">
            <TodoList
              deleteTodo={deleteTodo}
              toogleTodo={toogleTodo}
              todos={todos}
            />
          </div>
          <div className="col-md-5">
            <TodoForm addTodo={addTodo} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
