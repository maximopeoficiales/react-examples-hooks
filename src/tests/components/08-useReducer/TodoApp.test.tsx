import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import React from "react";
import TodoApp from "../../../components/08-useReducer/TodoApp/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";

describe("Pruebas en TodoApp", () => {
  Storage.prototype.setItem = jest.fn();
  const wrapper = shallow(<TodoApp />);
  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("debe de agregar un todo", () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      let addTodo: any = wrapper.find("TodoForm").prop("addTodo");
      addTodo(demoTodos[0]);
      addTodo(demoTodos[1]);
    });

    expect(wrapper.find("h1").text().trim()).toBe("TodoApp 2");
    // se agregaron dos todos por lo que se espera que se halla llamado el localstorage 2 veces
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test("debe eliminar un todo", () => {
    let addTodo: any = wrapper.find("TodoForm").prop("addTodo");
    let deleteTodo: any = wrapper.find("TodoList").prop("deleteTodo");
    addTodo(demoTodos[0]);
    deleteTodo(demoTodos[0]);
    expect(wrapper.find("h1").text().trim()).toBe("TodoApp 0");
  });
});
