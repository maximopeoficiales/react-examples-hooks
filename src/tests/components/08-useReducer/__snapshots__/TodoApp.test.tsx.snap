// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoApp debe mostrarse correctamente 1`] = `
<div
  className="row justify-content-center"
  data-testid="TodoApp"
>
  <div
    className="col-md-10"
  >
    <h1
      className="text-center font-weight-bold my-4"
    >
      TodoApp 
      0
    </h1>
    <div
      className="row"
    >
      <div
        className="col-md-7"
      >
        <TodoList
          deleteTodo={[Function]}
          todos={Array []}
          toogleTodo={[Function]}
        />
      </div>
      <div
        className="col-md-5"
      >
        <TodoForm
          addTodo={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
